proxy/src/fs_watch.rs

```diff
@@ -0,0 +1,701 @@
+
+        files.iter_mut().fold(watch, |watch, file| {
+            file.write_all(b"b").unwrap();
+            let (item, watch) = next_change(&mut rt, watch).unwrap();
+            assert!(item.is_some());
+            watch
+        });
+    }
+
+    #[cfg(not(target_os = "windows"))]
+    fn test_detects_modification_double_symlink(
+        fixture: Fixture,
+        watch: Watch<()>,
+        bg: Box<Future<Item = (), Error = ()>>
+    ) {
+        let Fixture { paths, dir, mut rt } = fixture;
+
+        let real_data_path = dir.path().join("real_data");
+        let data_path = dir.path().join("data");
+        fs::create_dir(&real_data_path).unwrap();
+        symlink(&real_data_path, &data_path).unwrap();
+
+        let mut files = paths.iter().map(|p| {
+            let path = real_data_path.clone()
+                .join(p.file_name().unwrap());
+            create_and_write(path, b"a").unwrap()
+        })
+        .collect::<Vec<_>>();
+
+        for path in &paths {
+            symlink(
+                data_path.clone().join(path.file_name().unwrap()),
+                path
+            ).unwrap();
+        }
+
+        rt.spawn(bg);
+
+        files.iter_mut().fold(watch, |watch, file| {
+            file.write_all(b"b").unwrap();
+            let (item, watch) = next_change(&mut rt, watch).unwrap();
+            assert!(item.is_some());
+            watch
+        });
+    }
+
+    #[cfg(not(target_os = "windows"))]
+    fn test_detects_double_symlink_retargeting(
+        fixture: Fixture,
+        watch: Watch<()>,
+        bg: Box<Future<Item = (), Error = ()>>
+    ) {
+        let Fixture { paths, dir, mut rt } = fixture;
+
+        let real_data_path = dir.path().join("real_data");
+        let real_data_path_2 = dir.path().join("real_data_2");
+        let data_path = dir.path().join("data");
+        fs::create_dir(&real_data_path).unwrap();
+        fs::create_dir(&real_data_path_2).unwrap();
+        symlink(&real_data_path, &data_path).unwrap();
+
+        // Create the first set of files
+        for path in &paths {
+            let path = real_data_path
+                .clone()
+                .join(path.file_name().unwrap());
+            create_and_write(path, b"a").unwrap();
+        }
+
+        // Symlink those files into `data_path`
+        for path in &paths {
+            let data_file_path = data_path
+                .clone()
+                .join(path.file_name().unwrap());
+            symlink(data_file_path, path).unwrap();
+        }
+
+        // Create the second set of files.
+        for path in &paths {
+            let path = real_data_path_2
+                .clone()
+                .join(path.file_name().unwrap());
+            create_and_write(path, b"b").unwrap();
+        }
+
+        rt.spawn(bg);
+
+        fs::remove_dir_all(&data_path).unwrap();
+        symlink(&real_data_path_2, &data_path).unwrap();
+
+        let (item, _) = next_change(&mut rt, watch).unwrap();
```

Comment:
Did you verify that if you comment out the two lines above this, then `next_change() ` will hang because there's no notifications? That is, did you verify that the `symlink()` call above is actually what causes the notification?

Reply:
Ah, you're right, this test still passes with those lines removed. Thanks for catching that!

 I've added a call to `next_change(...)` prior to creating the symlink, after which the test _will_ fail if we don't create the symlink.